test/RequestStreamTest_CancelAfterRequest.cpp

```diff
@@ -0,0 +1,157 @@
+
+#include <folly/io/async/ScopedEventBaseThread.h>
+#include <gtest/gtest.h>
+#include <random>
+#include <thread>
+
+#include "RSocketTests.h"
+#include "yarpl/Flowable.h"
+#include "yarpl/flowable/TestSubscriber.h"
+
+#include "yarpl/test_utils/Mocks.h"
+
+using namespace yarpl;
+using namespace yarpl::flowable;
+using namespace rsocket;
+using namespace rsocket::tests;
+using namespace rsocket::tests::client_server;
+
+using namespace yarpl::mocks;
+using namespace ::testing;
+
+struct Batons {
+  folly::Baton<> serverFinished;
+  folly::Baton<> clientFinished;
+
+  void reset() {
+    serverFinished.reset();
+    clientFinished.reset();
+  }
+};
+
+class RaceyRequestAfterCancelTestHandler : public rsocket::RSocketResponder {
+ private:
+  class RaceySubscription : public ::yarpl::flowable::Subscription {
+   public:
+    RaceySubscription(Reference<Subscriber<Payload>> s, Batons& batons)
+        : subscriber_(std::move(s)), batons_(batons) {}
+
+    void request(int64_t n) {
+      LOCKSTEP_DEBUG("SERVER: recieve request(" << n << ")");
+      for (int i = 0; i < n; i++) {
+        subscriber_->onNext(Payload("f"));
+      }
+    }
+
+    void cancel() {
+      LOCKSTEP_DEBUG("SERVER: recieve cancel(), sending onComplete()");
+      subscriber_->onComplete();
+      subscriber_ = nullptr;
+      this->batons_.serverFinished.post();
+    }
+
+   private:
+    Reference<Subscriber<Payload>> subscriber_;
+    Batons& batons_;
+  };
+
+ public:
+  RaceyRequestAfterCancelTestHandler(Batons& batons) : batons_(batons) {}
+
+  Reference<Flowable<Payload>> handleRequestStream(Payload p, StreamId)
+      override {
+    EXPECT_EQ(p.moveDataToString(), "initial");
+    return Flowables::fromPublisher<Payload>(
+        [this](Reference<flowable::Subscriber<Payload>> subscriber) {
+
+          LOCKSTEP_DEBUG("SERVER: sending onSubscribe()");
+          auto subscription = make_ref<RaceySubscription>(subscriber, batons_);
+          subscriber->onSubscribe(subscription);
+        });
+  }
+
+ private:
+  Batons& batons_;
+};
+
+TEST(RequestStreamTest, RaceyRequestAfterCancel) {
+  std::mt19937 rand(123); // we want a PRNG here
+  std::uniform_int_distribution<> dis_0_100(0, 100);
+
+  // number to request each request(n) call
+  auto const will_request = 100;
+
+  Batons batons;
+  folly::ScopedEventBaseThread worker;
+  auto server =
+      makeServer(std::make_shared<RaceyRequestAfterCancelTestHandler>(batons));
+  auto client = makeClient(worker.getEventBase(), *server->listeningPort());
+  auto requester = client->getRequester();
+
+  for (int i = 0; i < 1000; i++) {
+    auto at_response = 0;
+    Reference<Subscription> subscription;
+
+    auto subscriber = Subscribers::create<std::string>(
+        // onSubscribe
+        [&](Reference<Subscription> subscription_) {
+          subscription = subscription_;
+          subscription->request(will_request);
+
+          auto chance = dis_0_100(rand);
```

Comment:
Nit: Just use `folly::Random::rand32(100)`, much shorter, doesn't need the distribution and generator objects.

Reply:
We want something pseudorandom that we control the seed to, though. I couldn't find docs on folly::Random for how to do that :(

Reply:
Oh you're right, I didn't notice you were controlling the seed.  I don't see a way of doing that with `folly::Random` either :(